src/Interpreters/Cache/FileCache.cpp

```diff
@@ -946,6 +955,116 @@ bool FileCache::tryReserve(
+    try
+    {
+        /// Collect at most `keep_up_free_space_remove_batch` elements to evict,
+        /// (we use batches to make sure we do not block cache for too long,
+        /// by default the batch size is quite small).
+        limits_satisfied = main_priority->collectCandidatesForEviction(
+            desired_size, desired_elements_num, keep_up_free_space_remove_batch, stat, eviction_candidates, lock);
+
+#ifdef ABORT_ON_LOGICAL_ERROR
+        /// Let's make sure that we correctly processed the limits.
+        if (limits_satisfied && eviction_candidates.size() < keep_up_free_space_remove_batch)
+        {
+            const auto current_size = main_priority->getSize(lock);
+            chassert(current_size >= stat.total_stat.releasable_size);
+            chassert(!size_limit
+                        || current_size <= desired_size
+                        || current_size - stat.total_stat.releasable_size <= desired_size);
+
+            const auto current_elements_count = main_priority->getElementsCount(lock);
+            chassert(current_elements_count >= stat.total_stat.releasable_count);
+            chassert(!elements_limit
+                        || current_elements_count <= desired_elements_num
+                        || current_elements_count - stat.total_stat.releasable_count <= desired_elements_num);
+        }
+#endif
+
+        if (shutdown)
+            return;
+
+        if (eviction_candidates.size() > 0)
+        {
+            LOG_TRACE(log, "Current usage {}/{} in size, {}/{} in elements count "
+                    "(trying to keep size ration at {} and elements ratio at {}). "
+                    "Collected {} eviction candidates, "
+                    "skipped {} candidates while iterating",
+                    main_priority->getSize(lock), size_limit,
+                    main_priority->getElementsCount(lock), elements_limit,
+                    desired_size, desired_elements_num,
+                    eviction_candidates.size(), stat.total_stat.non_releasable_count);
+
+            lock.unlock();
+
+            /// Remove files from filesystem.
+            eviction_candidates.evict();
+
+            /// Take lock again to finalize eviction,
+            /// e.g. to update the in-memory state.
+            lock.lock();
+            eviction_candidates.finalize(nullptr, lock);
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+
+        if (eviction_candidates.size() > 0)
+            eviction_candidates.finalize(nullptr, lockCache());
+
+        /// Let's catch such cases in ci,
+        /// in general there should not be exceptions.
+        chassert(false);
+    }
+
+    LOG_TRACE(log, "Free space ratio keeping thread finished in {} ms", watch.elapsedMilliseconds());
+
+    if (limits_satisfied)
+        keep_up_free_space_ratio_task->scheduleAfter(general_reschedule_ms);
+    else
+        keep_up_free_space_ratio_task->schedule();
```

Comment:
I'm not sure if it's a realistic scenario: if a task fails to free up the space repetitively (`collectCandidatesForEviction` returns `false`), it will be re-scheduled for immediate processing. Can it stuck in this state?

Reply:
Yes, it is not a realistic scenario that we fail to free up required space because of not enough evictable/freeable cache entries, but `limits_satisfied` will also be `false` if we did not free up all the needed space because of the eviction batch size (by default it is quite small), so I thought it makes sense to schedule immediately in this case. I do not think it can get stuck in this state because this could possibly happen if cache constantly grows in size at a high speed, but in this case we should also get into lock contention case and get into another reschedule here: https://github.com/ClickHouse/ClickHouse/blob/3e4d9bf73abc072c2cb33b462cee5527212f71a1/src/Interpreters/Cache/FileCache.cpp#L969-L973
I will add profile event for the number of eviction tries, so we will be able to see if it is stuck in constant reschedule on the dashboard.